032-projects/proposal.md

```diff
@@ -0,0 +1,274 @@
+ci/resources/booklit-status.yml
+ci/tasks/test.yml
+ci/tasks/build.yml
+ci/pipelines/booklit.yml
+```
+
+## Anatomy of a `project.yml`
+
+A minimal project config has a `name` and a `plan`:
+
+```yaml
+name: ci
+plan: # ...
+```
+
+The `name` should be short and sweet - similar to a resource name. The project config directory will be available to every build executed within the project, as an artifact named after this value. ([Why only the config directory?](#exposing-only-the-config-directory))
+
+Concourse will continuously check the project resource for new versions and execute the project's `plan` when a new version is found. All pipelines, tasks, resources, and resource types defined in the project will be usable by the project's `plan`.
+
+Within the `plan`, you may choose to just run a task, using projects as a simple pipeline-less CI flow:
+
+```yaml
+name: ci
+plan:
+- get: booklit
+- task: test
+```
+
+...or you may choose to leverage the `set_pipeline` step to automate your pipeline configuration:
+
+```yaml
+name: ci
+plan:
+- set_pipeline: booklit
+```
+
+New project builds can be triggered by `get` steps with `trigger: true`, however `get` steps cannot specify any `passed` constraints:
+
+```yaml
+name: ci
+plan:
+- get: booklit
+- get: golang
+  trigger: true
+- task: test
+  image: golang
+```
+
+If your pipelines use templating, you can automate that just by using a `task`:
+
+```yaml
+name: ci
+plan:
+- task: generate-pipeline
+- set_pipeline: my-pipeline
+  file: generated-pipeline/pipeline.yml
+```
+
+This versatility should help smoothen Concourse's learning curve - there now exists a simpler CI/CD workflow suitable for side-projects and simple workflows that don't really need pipelines at all, and along the way we've enabled arbitrary automation needs for more complex workflows.
+
+## Project-level resources
+
+Projects define resources under the `resources/` directory. These resources can be used by all pipelines within the project.
+
+Resources defined at the project level can be propagated between pipelines using `passed` constraints. This is not possible with pipeline-local resource definitions.
+
+## Pipeline-local resources
+
+Pipelines can still define their own resources in addition to the project scope.
+
+There is some risk for confusion as developers will have two places to look for resource definitions, but the benefits seem to outweigh the risk:
+
+* Pipeline-local resources can be used to prevent polluting the project-wide resource scope with resources that are only ever useful for a particularly specialized pipeline.
+* Pipeline-local resources are natural fit for pipeline templates which use `((vars))` to configure their resources. For example, a branch pipeline template would have a pipeline-local resource definition for its branch.
+
+Concourse will prevent 'resource shadowing' - any pipeline which defines a local resource matching the name of a project-wide resource should error upon configuration.
```

Comment:
I don't have any formed opinion on this, but I'm curious if there's reasoning/past examples behind why Concourse will prevent resource shadowing?

Reply:
I'm just being cautious. I don't really have any past examples as right now pipelines just have a flat resource namespace.

This is the first time adding two levels of scope. I'm usually not a fan of having to look in multiple places for things that go into my CI. GoCD for example lets you set environment variables at multiple levels, so when I'm trying to find where something is set I have to traverse the whole tree.

In this case there seem to be pretty solid justifications for allowing both levels, so I want to mitigate any potential for confusion I can before we paint ourselves into a corner. To me shadowing resource definitions feels just as risky as shadowing variables in a programming language, and lots of languages emit warnings or errors for it. So until someone provides a really solid use case for shadowing vs. just giving one resource a different name or making all the definitions local to each pipeline instead, I think we should prevent it.

Here's a made up example: say my project uses Go 1.12 and defines a project-wide `golang` resource for it. I only intend to support one version of Go at a time, so the name is just `golang` instead of `golang-1.12` as I don't want to have to update all my pipelines every time I bump the version.

Next I bring in a pipeline that already had its own `golang` resource. This one is also set to version 1.12, but it'd be easy to forget about it. A new version of Go comes out, I bump my project to 1.13, but that pipeline's stays at 1.12 and all the while it just says `golang`. It could take a while to notice that kind of thing.